lib/ConversionProvider.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Pandoc;
+
+use OCP\Files\Conversion\ConversionMimeTuple;
+use OCP\Files\Conversion\IConversionProvider;
+use OCP\Files\File;
+use OCP\IL10N;
+use RuntimeException;
+
+class ConversionProvider implements IConversionProvider {
+	private ?array $types = null;
+
+	public function __construct(
+		private IL10N $l10n,
+	) {
+	}
+
+	public function getSupportedMimeTypes(): array {
+		$types = [];
+		foreach ($this->getPandocTypes() as $mimeType => $typeData) {
+			$types[] =
+				new ConversionMimeTuple(
+					$mimeType,
+					array_values(array_filter($this->getPandocTypes(), function ($mimeType) use ($typeData) {
+						return $typeData['mime'] !== $mimeType;
+					}))
+				);
+		}
+		return $types;
+	}
+
+	public function convertFile(File $file, string $targetMimeType): mixed {
+		$fileContent = $file->getContent();
+
+		$from = $this->getPandocForMime($file->getMimeType());
+		$to = $this->getPandocForMime($targetMimeType);
+
+		if ($from === null || $to === null) {
+			throw new RuntimeException('Invalid source or target file format for pandoc conversion');
+		}
+
+		$pandoc = new \Pandoc\Pandoc;
+		return $pandoc
+			->input($fileContent)
+			->option('from', $from)
+			->option('to', $to)
+			->noStandalone()
+			->execute();
+	}
+
+	private function getPandocTypes(): array {
+		// FIXME pdf should be optional (->option('pdf-engine', '/usr/bin/pdflatex'))
+		if ($this->types === null) {
+			$this->types = [
+				'text/markdown' => [
+					'mime' => 'text/markdown',
+					'pandoc' => 'gfm',
+					'extension' => 'md',
+					'name' => $this->l10n->t('Markdown (.md)'),
+				],
+				'application/vnd.openxmlformats-officedocument.wordprocessingml.document' => [
+					'mime' => 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+					'pandoc' => 'docx',
+					'extension' => 'docx',
+					'name' => $this->l10n->t('Document (.docx)'),
+				],
+				'application/pdf' => [
+					'mime' => 'application/pdf',
+					'pandoc' => 'pdf',
+					'extension' => 'pdf',
+					'name' => $this->l10n->t('PDF (.pdf)'),
+				],
+				'application/msword' => [
+					'mime' => 'application/msword',
+					'pandoc' => 'doc',
+					'extension' => 'doc',
+					'name' => $this->l10n->t('Word (.doc)'),
```

Comment:
Extension needs to be checked with upstream PR discussion
https://github.com/nextcloud/server/pull/49922/files#r1917974284